### ISSUE TITLE
Feature: form validate in front-end 

### ISSUE THREAD

**Issue by terryatgithub** [Instawork/hyperview#305]
hi @adamstep, do you consider how to do form validate totally in front-end?  not through sever side ?
because in our app, we need do a lot of actions of user info fill and submit, and the function of front-end form validation is vital to us.
the PR #304 is currently the method we thought of.
we'd love to hear your comment and opinions very much. 
thank you.

**Comment by adamstep** [comment]
Thanks for bringing up this topic, I've added it to the roadmap for discussion.

Your demo shows a few use cases and requirements for client-side validation:
1. trigger per-field validation on input change, or on blur
2. based on valid/invalid state, we should show/hide certain elements or error messages
3. based on valid/invalid state of all fields in the form, enable/disable certain elements on the page (like submit button)

Your demo also included an example of linked fields for sums. (Sum should be 100, and values populate in other fields). I think this is a separate issue from client-side validation, so perhaps we should not include that in the current discussion.

For **1**, we already support triggering a behavior on [blur](https://hyperview.org/docs/reference_behavior_attributes#blur) and [focus](https://hyperview.org/docs/reference_behavior_attributes#focus). There is also an issue to trigger behaviors on changes to an input field: #298 .

With a way to trigger behaviors on blur and change, we can focus on requirement **2**. I'm taking some inspiration from the [ParselyJS library](https://parsleyjs.org/doc/index.html#validators) for declarative form validation on the web. Here's an example of what it could look like:

```xml
<doc
  xmlns="https://hyperview.org/hyperview"
  xmlns:v="https://hyperview.org/hyperview-validation"
>
  <screen>
    <body>
      
      <text-field name="age">
        <behavior trigger="blur" action="validate">
          <v:required message="Age is required." />
          <v:integer message="Please enter a number" />
          <v:range min="18" max="99" message="Age must be between 18 and 99" />
        </behavior>

        <behavior trigger="invalid" action="show-invalid-message" target="#name-error" />
        <behavior trigger="valid" action="hide" />
      </text-field>

      <text id="name-error" hide="true"></text>

    </body>
  </screen>
</doc> 
``` 

Some new features include:
- XML namespace for `https://hyperview.org/hyperview-validation`
- `validate` behavior action. This action will run validation on the text field.
- `<v:required />` and other elements that specify how to validate the field. They can specify custom error messages.
- `invalid` and `valid` behavior triggers. After validation, one of these behaviors is triggered based on whether the field is valid or not.
- `show-invalid-message` action. Inserts the error message text into the given target element on the page.

Perhaps Hyperview could ship with some built-in validators. But it would be possible to add custom validators in the same way we support custom behaviors and components. This would allow custom JS code to run to perform validation.

For **3**, we could support the same `validate` action and `invalid` and `valid` triggers on the `<form>` element itself. This would support enabling/disabling the submit button based on the state of all fields in the form.


@flochtililoch , @ashwinibm , @terryatgithub Please leave comments and suggestions based on this proposal. Hopefully it gets the conversation going in terms of how to support validation in a generic and extensible way.

**Comment by terryatgithub** [comment]
hi @adamstep , it's amazing you move so fast. 
I've read the Parsley documents roughly, and it's powerful and supports common basic types validation, custom validator and asynchronous validation. and it uses a jQuery-like syntax and directly operates on DOM.

I also studied the form validation method of the most popular Open Source UI component libraries in China:
- [AntD form](https://ant.design/components/form/) based on `React`
- [Element-UI form](https://element.eleme.cn/#/en-US/component/form) based on `Vue`
   both uses [async-validator](https://github.com/yiminghe/async-validator) as its asynchrous validation library.

Here is an example of what it look like in `AntD form`: 
it simply adds an attribute `rules`, and contains all validation logic inside it.

```Javascript
  return (
    <Form
      {...layout}
      name="basic"
      initialValues={{ remember: true }}
      onFinish={onFinish}
      onFinishFailed={onFinishFailed}
    >
      <Form.Item
        label="Username"
        name="username"
        rules={[{ required: true, message: 'Please input your username!' }]}
      >
        <Input />
      </Form.Item>

      <Form.Item
        label="Password"
        name="password"
        rules={[{ required: true, message: 'Please input your password!' }]}
      >
        <Input.Password />
      </Form.Item>

    </Form>
  );
};
```
it also supports custom validator function:
```javascript
<Form.Item
        name="confirm"
        label="Confirm Password"
        dependencies={['password']}
        hasFeedback
        rules={[
          {
            required: true,
            message: 'Please confirm your password!',
          },
          ({ getFieldValue }) => ({
            validator(_, value) {
              if (!value || getFieldValue('password') === value) {
                return Promise.resolve();
              }
              return Promise.reject(new Error('The two passwords that you entered do not match!'));
            },
          }),
        ]}
      >
        <Input.Password />
      </Form.Item>
```

This syntax looks simpler and more closer to the React JSX syntax. 
Is it possible to implement it in this way ?

**Comment by adamstep** [comment]
Thanks @terryatgithub for the suggestion. In general, one of our goals with the design of HV is to avoid adding scripts and code to the XML markup. We want the custom code live within custom components and behaviors, that way the XML markup can be more declarative. This helps with aspects such as server-side templating (it would be harder to templatize your script tag to customize the validation states).

The downside of my proposal is that adding a completely custom validation requires an RN app update, so it is not as flexible as your proposal. However, it would be possible to create a generic custom validator with this API:

```xml
<doc
  xmlns="https://hyperview.org/hyperview"
  xmlns:v="https://hyperview.org/hyperview-validation"
  xmlns:cv="https://mycompany.com/custom-validation"
>
  <screen>
    <body>
      
      <text-field name="age">
        <behavior trigger="blur" action="validate">
          <v:required message="Age is required." />
          <cv:custom-validation>
          <![CDATA[
            (value, callback) => { /* custom code here similar to ElementUI */ }
          ]]>  
          </cv:custom-validation>
        </behavior>

        <behavior trigger="invalid" action="show-invalid-message" target="#name-error" />
        <behavior trigger="valid" action="hide" />
      </text-field>

      <text id="name-error" hide="true"></text>

    </body>
  </screen>
</doc> 
```

This way, you can implement `cv-custom-validation` one time, and use it as a generic validator that can run any code to reach an answer. What do you think?

**Comment by adamstep** [comment]
Depends on #298.

**Comment by eazuka** [comment]
Hello everyone and thanks to the instawork team, @adamstep for the really nice project and making it open source. I am about starting a project with hyperview which is why i'm here. @adamstep has any progress been made on this client side validation proposal as discussed here?

**Comment by adamstep** [comment]
@eazuka we have moved the discussion here: https://github.com/Instawork/hyperview/discussions/332

I'm hoping to nail down the initial API soon and start the implementation phase.
